pkg/distributors/contracts/MultiRewards.sol

```diff
@@ -0,0 +1,284 @@
+    }
+
+    /* ========== MUTATIVE FUNCTIONS ========== */
+
+    function setRewardsDistributor(address _rewardsToken, address _rewardsDistributor) external onlyOwner {
+        rewardData[_rewardsToken].rewardsDistributor = _rewardsDistributor;
+    }
+
+    function stake(uint256 amount) external {
+        stake(amount, msg.sender);
+    }
+
+    function stake(uint256 amount, address receiver) public nonReentrant whenNotPaused updateReward(receiver) {
+        require(amount > 0, "Cannot stake 0");
+        _totalSupply = _totalSupply.add(amount);
+        _balances[receiver] = _balances[receiver].add(amount);
+        stakingToken.safeTransferFrom(msg.sender, address(this), amount);
+        emit Staked(receiver, amount);
+    }
+
+    /**
+     * @notice Stake tokens using a permit signature for approval
+     * @param amount    Amount of allowance
+     * @param deadline  The time at which this expires (unix time)
+     * @param v         v of the signature
+     * @param r         r of the signature
+     * @param s         s of the signature
+     */
+    function stakeWithPermit(
+        uint256 amount,
+        uint256 deadline,
+        uint8 v,
+        bytes32 r,
+        bytes32 s
+    ) public {
+        IERC20Permit(address(stakingToken)).permit(msg.sender, address(this), amount, deadline, v, r, s);
+        stake(amount, msg.sender);
+    }
+
+    function withdraw(uint256 amount) public nonReentrant updateReward(msg.sender) {
+        require(amount > 0, "Cannot withdraw 0");
+        _totalSupply = _totalSupply.sub(amount);
+        _balances[msg.sender] = _balances[msg.sender].sub(amount);
+        stakingToken.safeTransfer(msg.sender, amount);
+        emit Withdrawn(msg.sender, amount);
+    }
+
+    /**
+     * @notice Allows a user to claim any rewards which are due.
+     */
+    function getReward() public nonReentrant updateReward(msg.sender) {
+        for (uint256 i; i < rewardTokens.length; i++) {
+            address _rewardsToken = rewardTokens[i];
+            uint256 reward = rewards[msg.sender][_rewardsToken];
+            if (reward > 0) {
+                rewards[msg.sender][_rewardsToken] = 0;
+                IERC20(_rewardsToken).safeTransfer(msg.sender, reward);
+                emit RewardPaid(msg.sender, _rewardsToken, reward);
+            }
+        }
+    }
+
+    /**
+     * @notice Allows a user to claim any rewards to internal balance
+     */
+    function getRewardAsInternalBalance() public nonReentrant updateReward(msg.sender) {
+        IVault.UserBalanceOp[] memory ops = new IVault.UserBalanceOp[](rewardTokens.length);
+
+        for (uint256 i; i < rewardTokens.length; i++) {
+            address _rewardsToken = rewardTokens[i];
+            uint256 reward = rewards[msg.sender][_rewardsToken];
+
+            if (reward > 0) {
+                rewards[msg.sender][_rewardsToken] = 0;
+
+                ops[i] = IVault.UserBalanceOp({
```

Comment:
@nventuro does the vault handle empty `UserBalanceOp`s?

(If there were multiple reward tokens and a user wasn't receiving a reward for all of them)

Reply:
That should end up being equivalent to 
```
op = {
  asset: address(0),
  amount: 0,
  sender: address(0),
  recipient: address(0),
  kind: DEPOSIT_INTERNAL
}
```

Which would end up falling afoul of the relayer auth checks. It's probably worth just doing the zero value transfers instead of looping over the assets twice.

Reply:
This should only really happen in situations where a rewards token is no longer being fed into the staking contract.

It might be worth maintaining separate arrays of assets - one for all rewards ever and one for "active" rewards. An admin can then set tokens inactive after the rewards period runs out and then users can explicitly claim them. This could be automated to not use an admin but that's implementation details.

Reply:
Just noticed that we're doing deposits here rather than internal transfers. I think in this case it's worth looping twice to avoid the external calls to each token contract.